diff --git a/answers-traps.txt b/answers-traps.txt
new file mode 100644
index 0000000..e608f67
--- /dev/null
+++ b/answers-traps.txt
@@ -0,0 +1,6 @@
+Q1: a2
+Q2: The compiler inlined f and g into main.
+Q3: 0x630
+Q4: 0x38
+Q5: HE110 World. i should be 0x726c6400. 57616 doesn't change.
+Q6: It prints a garbage value (or previous value) from register a2 because printf expects an argument there but none was provided.
\ No newline at end of file
diff --git a/gradelib.py b/gradelib.py
index f0d4934..21dd2bc 100644
--- a/gradelib.py
+++ b/gradelib.py
@@ -238,7 +238,7 @@ def make(*target):
     post_make()
 
 def show_command(cmd):
-    from pipes import quote
+    from shlex import quote
     print("\n$", " ".join(map(quote, cmd)))
 
 def maybe_unlink(*paths):
diff --git a/kernel/defs.h b/kernel/defs.h
index a3c962b..4b82fc2 100644
--- a/kernel/defs.h
+++ b/kernel/defs.h
@@ -80,6 +80,7 @@ int             pipewrite(struct pipe*, uint64, int);
 void            printf(char*, ...);
 void            panic(char*) __attribute__((noreturn));
 void            printfinit(void);
+void            backtrace(void);
 
 // proc.c
 int             cpuid(void);
diff --git a/kernel/printf.c b/kernel/printf.c
index 1a50203..90e3db6 100644
--- a/kernel/printf.c
+++ b/kernel/printf.c
@@ -119,6 +119,7 @@ void
 panic(char *s)
 {
   pr.locking = 0;
+  backtrace();
   printf("panic: ");
   printf(s);
   printf("\n");
@@ -127,6 +128,25 @@ panic(char *s)
     ;
 }
 
+void
+backtrace(void)
+{
+  uint64 fp = r_fp();
+  uint64 stack_bottom = PGROUNDDOWN(fp);  // đáy của kernel stack page
+
+  printf("backtrace:\n");
+
+  while (fp >= stack_bottom + 16 && fp < stack_bottom + PGSIZE) {
+    uint64 ra = *(uint64*)(fp - 8);        // return address
+    printf("%p\n", ra);            
+
+    uint64 prev_fp = *(uint64*)(fp - 16);  // saved fp
+    if (prev_fp <= fp || prev_fp >= stack_bottom + PGSIZE)
+      break;
+    fp = prev_fp;
+  }
+}
+
 void
 printfinit(void)
 {
diff --git a/kernel/riscv.h b/kernel/riscv.h
index 20a01db..90b522b 100644
--- a/kernel/riscv.h
+++ b/kernel/riscv.h
@@ -330,6 +330,14 @@ sfence_vma()
 typedef uint64 pte_t;
 typedef uint64 *pagetable_t; // 512 PTEs
 
+static inline uint64
+r_fp()
+{
+  uint64 x;
+  asm volatile("mv %0, s0" : "=r" (x));
+  return x;
+}
+
 #endif // __ASSEMBLER__
 
 #define PGSIZE 4096 // bytes per page
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 3b4d5bd..4ef466f 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -51,6 +51,7 @@ sys_sbrk(void)
 uint64
 sys_sleep(void)
 {
+  backtrace();
   int n;
   uint ticks0;
 
